keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,348 @@
+
+The developer also would not need to worry about feature compatibility when there is a mixed version cluster.
+
+#### Story 2
+
+Users of etcd cluster would be able to find out which features are enabled in the etcd cluster and decide on how to use them downstream. 
+
+Currently Kubernetes uses [`FeatureSupportChecker`](https://github.com/kubernetes/kubernetes/blob/db82fd1604ebf327ab74cde0a7158a8d95d46202/staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature/feature_support_checker.go#L42) to check if a feature is supported by etcd, which basically checks the etcd version and compares the version with a hard-coded map of feature availability for different etcd versions.
+
+#### Story 3
+
+In a HA cluster, users should be able to enable/disable a feature without bring down the whole cluster, by restarting the servers with the new feature gate flag one by one, and the feature enablement would be consistent across all members. 
+
+#### Story 4
+
+During cluster upgrade/downgrade, feature changes across versions should have predictable behavior as allowed by the cluster version. 
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Could it done on the client side instead of inside etcd? 
+One could argue that it is easy to know if a feature enabled for the whole cluster by querying each member if the feature is enabled on the client side. But there are several caveates with this approach:
+* not all members are available all the time. if some member is not available, it is questionable if the feature can be used, and that could break the HA assumption. 
+* some feature might change how a raft message is sent and applied, and the order of messages relative to the index the feature is enabled is critical. Here the consistent index is more important than the time, and it is hard to get that outside etcd. 
+
+#### Should we prevent a new member with an inconsistent feature gate from joining the cluster?
+
+One way to prevent inconsistencies of feature gate between different members is to prevent a new member with an inconsistent feature gate from joining the cluster in the first place. 
+We do not think this is the right way because: 
+* we want to keep the ability to turn on/off a feature in a rolling sequence without bringing down the whole cluster. 
+
+
+### Risks and Mitigations
+
+#### Backport Risks
+
+To support feature gate as soon as in 3.6, we are backporting some proto changes into 3.5. There is the potential risk of changing a stable release. But we do not think this is a real risk, because it does not involve any changes other than adding a new proto field in an existing proto, and 3.5 server does not write or use this new field. Protos are inherently backward compatible. This change should also not affect upgrade/downgrade of 3.5.
```

Comment:
I think it's also worthwhile to highlight the benefit of backporting. Since proto3 fields are optional (and unknown fields are ignored), we want to provide rationale for backporting this change.

Reply:
actually because we are introducing a new raft message type, we have to backport. Changed the paragraph.